### ISSUE TITLE
JSON:API

### ISSUE THREAD

**Issue by ThexXTURBOXx** [tionis/studip-sync#1]
Just FYI: Stud.IP 5.4 will introduce the new JSON:API, which differs quite a lot from the current REST API.
See here for the documentation: https://docs.gitlab.studip.de/entwicklung/docs/jsonapi/routen
From Stud.IP 6.0 on, the REST API will be completely removed and only the JSON:API will persist.
However, you will most probably have enough time to migrate anyway because the University of Passau usually takes a quite a lot of time to update to new versions.

Edit: My app StudiPassau actually migrated not too long ago - in case you need some "inspiration":
https://github.com/ThexXTURBOXx/studipassau/blob/main/lib/bloc/cubits/files_cubit.dart
https://github.com/ThexXTURBOXx/studipassau/blob/main/lib/bloc/repos/files_repo.dart

**Comment by tionis** [comment]
Thanks for the information, I think I'll put it into the README.  
I originally wrote this script in janet a few years ago for myself only and only translated it now to share it to a few friends.  
I'll probably have finished my masters until the university finishes the migration :grin: 

Do you know any better way to authenticate a cli application to the studip backend other than cookie loading or registering some oAuth application with the admins (which would be kinda weird for a small cli script)?

**Comment by ThexXTURBOXx** [comment]
You're more than welcome :)
And yes, indeed, that could very well happen!

The way authentication initially worked in StudiPassau (before the big Flutter rewrite) was by just simulating the usual Shibboleth flow: https://github.com/ThexXTURBOXx/shib-uni-passau/blob/master/src/main/java/de/femtopedia/studip/shib/ShibbolethClient.java
It's not a very nice way to do it, but it worked back then (maybe even still works?!)

**Comment by tionis** [comment]
Ah, I think we talked about this a few years ago :smiley: 
The shibboleth simulation flow is pretty cursed though, I think I'll stick to loading cookies and maybe look into the json API when I find the motivation.

**Comment by ThexXTURBOXx** [comment]
Yes, at some point in 2018/19 we had talked about it :D
I would also suggest not doing it this way and just stay with cookies.
There is also this project (which tries to do the same as your project right here): https://github.com/yannik995/StudIP-Downloader
But it actually just lets the user manually input the cookie without trying to load it from somewhere

**Comment by tionis** [comment]
Yeah I've looked at the project in the past, but it's in C# ...  

My project is very narrow and focused in scope for my workflow, but I just added some config options to make it easier to change the auth strategy.
